### ISSUE TITLE
Nullness EEA

### ISSUE THREAD

**Issue by maggu2810** [openhab/openhab-core#888]
<a alt="bounty" href="https://www.bountysource.com/issues/75826293"><img src="https://api.bountysource.com/badge/issue?issue_id=75826293"></a>

I would like to resume the discussion about EEA for the nullness information of third party libraries (and the JRE files).

As we moved to pure Maven projects the situation changed a little bit.

First we need something that provides the EEAs we would like to consume (the approach here will not work for openHAB Core only but for all projects we would like to use it).

This is in my example a simple Maven project that contains the EEAs in the resources, so they are part of the created JAR file.

For the normal Maven build (command line) of our openHAB projects it is enough to change a simple setting of the used compiler that instruct the compiler to use EEAs that are part of the current classpath. Additionally we need to add a the EEAs to the classpath that is fulfilled by adding an artifact that contains the EEAs to the dependencies.
See https://github.com/maggu2810/openhab-core/commit/6e26af1c8c201e1f9bd26fe7fc94af3371fbfbaa#diff-600376dffeb79835ede4a0b285078036R303 and ignore the additional property (not part of that integration).

For the command line all is done now.

So, what needs to be done to get the support in the IDE?
We need to inform the IDE about the EEAs we want to use.
I don't know about a way to state "use the classpath", but we can refer to a project that is part of the workspace. This would be the preffered solution for me.
If the project containing the EEAs will be part of every workspace the remaining task will be easy. Would this be acceptable if we add it to the setup file and e.g. place the JAR containing the EEAs to openhab-distro or another separate repo that will be included for every IDE setup?
To add the respective annotationpath to the classpath for Eclipse, we can use the property above (in the POM file) and a simple Maven plugin of LastNPE (it will use the value of the specific property and add it to the projects settings).

I have already done all the necessary changed for openHAB Core.

WDYT?

**Comment by maggu2810** [comment]
As there seems to be no interest of the core maintainers I close this one.

**Comment by wborn** [comment]
I'd love to see EEAs in place to get rid of all those false positive warnings for `Map`s etc. I just didn't have time to further investigate this subject so I could add a meaningful comment here. Personally I'd rather first see the IDE/M2 woes resolved.

**Comment by kaikreuzer** [comment]
Same for me. Interest is definitely there, but it's not on my top priority list atm. I wouldn't mind to keep this issue open nonetheless as it is a valid discussion.

**Comment by wborn** [comment]
So let's reopen the issue because we do have interest in using EEA.

It would also properly address the `Map.get(..)` false positives that many run into:
* https://github.com/eclipse/smarthome/issues/5908
* [community post](https://community.openhab.org/t/nonnullbydefault-and-hashmap/79845?u=wborn)

**Comment by wborn** [comment]
I've placed a [bounty](https://www.bountysource.com/issues/75826293-nullness-eea) on this issue!

**Comment by lujop** [comment]
For me no need of a bounty, I will be happy to work on this if it's defined and bounded.
As I commented in a [forum post](https://community.openhab.org/t/how-do-you-deal-with-jdt-static-null-analysis-limitations/95036/16) it doesn't seem a lot of work for me.

I propose my implementation idea, and if a maintainer can confirm he likes it, I start working on it:

Phase 1: 
I would start in a first PR be just using EEA for JDK API:

* Create a new openhab-eea project that starts with current lastnpe definitions for JDK API. And add more for new methods like Map.getOrElse, Optional.orElse, ...
* Integrate that new project in MVN in Openhab addons and core like [I did in this commit](https://github.com/lujop/openhab-addons/commit/c6d57b54f92893eced7be136828782ea8b13163f) 

I can confirm that it works in pure maven builds. And for Eclipse, I'm not an expert but it's very easy to configure for JDK manually:

![image](https://user-images.githubusercontent.com/920260/79020056-1792a580-7b78-11ea-9c73-0917efe5f95d.png)

Phase 2:
Start adding more EEA definitions to the openhab-eea for main libraries

**Comment by wborn** [comment]
Just reinvest the bounty into an issue you think is important. That's why I added my earned bounty here. :wink:

If I understand correctly there is a LastNPE Eclipse plugin that may help configuring Eclipse. 

The whole openHAB Eclipse configuration is automated using this Oomph setup file: 

https://github.com/openhab/openhab-distro/blob/master/launch/openHAB2.setup

For me it is important that both the Maven build and Eclipse IDE use the EEAs without developers having to do any manual configurations.

**Comment by lujop** [comment]
I will take a look for M2E configuration. If it works, the described approach will be ok?

**Comment by wborn** [comment]
The reason for this `openhab-eea` project is probably because lastnpe hasn't published any new artifacts to https://repo1.maven.org/maven2/org/lastnpe/eea/ since 2016 and we may want to add our own EEA (for less mainstream 3rd party libraries) that may not get accepted by lastnpe?

Do you know if it is possible to use the EEA in both Eclipse and Maven builds by just adding EEA Maven artifact dependencies to POMs? I think that would be the preferred way.

When the EEA work for core/add-ons, we probably also want to start using EEA in projects like [static-code-analysis](https://github.com/openhab/static-code-analysis). So if we can make the EEA work based fully on Maven dependencies, it is probably preferable to create our own GitHub repo with Jenkins build for publishing our EEA artifacts.

**Comment by lujop** [comment]
> The reason for this `openhab-eea` project is probably because lastnpe hasn't published any new artifacts to https://repo1.maven.org/maven2/org/lastnpe/eea/ since 2016 and we may want to add our own EEA (for less mainstream 3rd party libraries) that may not get accepted by lastnpe?

Yes, it's only to have a place to be agile to create our eea definitions and publish to mvn respository as jar.
 
> Do you know if it is possible to use the EEA in both Eclipse and Maven builds by just adding EEA Maven artifact dependencies to POMs? I think that would be the preferred way.

I hope so. For MVN build and other IDEs that use a more pure MVN based build (all except Eclipse?) it seems possibly by the test I did in the commit I commented in this thread.
The problem can be for Eclipse that needs to interpret the mvn file and be able to configure external annotations in its internal project configuration for eea.
This seems not possible by default, but to my understanding, that is what it does the m2e lastnpe plugin.
But it will need to be tested, because I make a quick look at the plugin at it points to an example project that, at least, at first sight, it doesn't seem to use the plugin.

>When the EEA work for core/add-ons, we probably also want to start using EEA in projects like static-code-analysis. So if we can make the EEA work based fully on Maven dependencies, it is probably preferable to create our own GitHub repo with Jenkins build for publishing our EEA artifacts.

That's my intention. If I'm not able to do that way I will stop on this.

**Comment by lujop** [comment]
Just a little update.
I've made some testing with lastnpe m2e plugin, and if setting appropriate properties (*m2e.jdt.annotationpath.**) they are recognized by Eclipse and project EEA locations are updated.

The problem I still see is that the values I can set from POM seem to be paths and don't know how to put an artifact maven coordinates.
I've opened [a question in LastNPE repository](https://github.com/lastnpe/eclipse-external-annotations-m2e-plugin/issues/40) to see if it's possible.
 
If it' not, a possible workaround will be to document that user must have defined a value for M2_REPO variable, and use something like ${M2_REPO}/org/openhab/1.0.0/openhab-eea-1.0.0.jar as M2E property value

**Comment by wborn** [comment]
Looks like you can get hold of the repo location in POMs using the following Maven expression: `${settings.localRepository}`, I don't know if the plugin also resolves such expressions?

See: https://mkyong.com/maven/where-is-maven-local-repository/

We also need to make sure the artifact is actually downloaded there.

**Comment by lujop** [comment]
For MVN it's working well, but I been not able to make it works for Eclipse.
I explain the situation and tests done to help if someone can continue with it.
For me, I've little knowledge on Eclipse (I'm not even a user) and it's not motivating to work on that in free time.

Look at [this difference to see the changes](https://github.com/lujop/openhab-addons/compare/influxdb2clean...lujop:test_eea?expand=1)

## Steps to make it work

### Install openhab-eea project
Clone and install https://github.com/lujop/openhab-eea
It contains original lastnpe definitions and two modifications:
added getOrDefault for Map
Defined *org.apache.commons.lang.StringUtils.lowerCase* parameter as not nullable (only for testing)

### Checkout branch test_eea
Branch https://github.com/lujop/openhab-addons/tree/test_eea
If you do a MVN build of influxdb addon you must see:
- Map.getOrDefault from InfluxDBConfiguration constructor don't give any warnings.
- ItemToStorePointCreator.calculateMeasurementName(Item, String) gives a compile error for passing null to StringUtils.lowerCase  (unreal test, just to have an example from an external library)

### For eclipse
- Install [lastnpe m2e maven](https://github.com/lastnpe/eclipse-external-annotations-m2e-plugin) plugin using that update site: http://www.lastnpe.org/eclipse-external-annotations-m2e-plugin-p2-site/
- Modify [main pom](https://github.com/lujop/openhab-addons/blob/test_eea/pom.xml):
Change *m2e.eea.annotationpath.jre* and *m2e.eea.annotationpath.maven* properties changing $MAVEN_REPO for your path to m2 repo.
- Make a mvn install
- Make a Eclipse MAven Update project

### What's missing
Make it work without having to modify the properties to point to your maven repo.
I've tried several alternatives:
- Use the maven dependency plugin to set the path of the dependency. Look most voted response at[ this SO thread.](https://stackoverflow.com/questions/2359872/can-i-use-the-path-to-a-maven-dependency-as-a-property/6934552#6934552)
- Simply put a environment MAVEN_REPO that user must configure in it's environment.

But none of them work.
The first possibly because [M2E update doesn't seem to work maven]([url](https://stackoverflow.com/questions/53042742/can-i-execute-maven-plugin-whith-eclipse-m2e-update/53042949#53042949)), at least completely.
And the second because Eclipse doesn't seem to like to have variables in EEA path. In fact if you try to use one manually in IDE they aren't used.

**Comment by wborn** [comment]
Thanks for the write-up and looking into it @lujop!
I hope the missing parts will be solved one day.

**Comment by wborn** [comment]
I think we have made enough progress to close this issue and award the bounty to you @J-N-K. :+1:

We can further improve it in follow up PRs (upgrade EEA dependency version, use EEAs with more projects, etc.)
